### PR TITLE
Configurability using config file

### PR DESCRIPTION
**Pull request by Victor-Schwan**
BEGINRELEASENOTES
- Paths to personal and detector-specific files/folders outsourced into the config file
- Choose config file with `argparse`
- Path operations using `pathlib`
- so far, only tested with `condorJobs_sim.py`

ENDRELEASENOTES

### CODE REVIEW

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py]: this looks familiar, @tmadlener do we have this already importable from somewhere in key4hep and if not should we make that happen?

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py]: I know it was already `testmatch` before but does it really take that long? I think we might want to aim for job sizes closer to `longlunch`

We also want to be able to set this from the config.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py]: This check looks like we only rewrite the files when we change the number of events per job? So if we just change the type of particles they are not written?

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (tmadlener) [reply TrackingPerformance/Condor/condorJobs_sim.py]: We were discussing putting [this bit](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/py_utils.py#L9-L49) into the k4FWCore utils, but ILDReconstruction was the only use case at the time. We could move it to k4FWCore.utils

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py]: Does something break when commenting it in? Without it the file is very useless :)

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py]: Is this needed (or even correct)? Before we were cloning `CLDConfig` here to have the `cld_steer.py` file available.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py:212]: good idea :)
or maybe even get rid of this or make it optional so that one can inspect the file before submission.
I think we can expect a user to be able to still type this one line by hand. If we want to be extra nice we could even print a small message from which this is directly copy-pasteable.

**Review by Zehvogel — COMMENTED** [review]
Thx! 

Some comments below. I did not look much at the other files yet.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: I only changed the Path. Maybe @gaswk can comment on this?

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: It is not entirely the same. We could tackle this in a later refactoring

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: Fair point. Done

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py:212]: Or we could add a `no-submit` flag to `argparse`?

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [reply TrackingPerformance/Condor/condorJobs_sim.py]: You would have "only changed the path" if this was `github.com/gaswk/FullSim` but it was `github.com/gaswk/CLDConfig`

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [reply TrackingPerformance/Condor/condorJobs_sim.py]: Fine with me. I just wanted to avoid having to many copies of the same thing flying around. :)

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [reply TrackingPerformance/Condor/condorJobs_sim.py:212]: yes a flag for this would also be a good idea but I am unsure if the default should be to submit it. :shrug:

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (nitasad) [reply TrackingPerformance/Condor/condorJobs_sim.py]: `CLDConfig` is required to use `cld_steer.py`. I use my version (`https://github.com/gaswk/CLDConfig.git`) because it includes configurable values for vertex resolution. However, it is preferable to use the official one `https://github.com/key4hep/CLDConfig`.

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (tmadlener) [inline TrackingPerformance/Condor/condorJobs_reco.py]: Do you want to keep the cloning in the batch jobs? Talking mostly about ILDConfig now, but these can become quite large and cloning can take quite a bit. If it's at all possible to lift the cloning to a "general setup step" that happens before the submission, etc. happens, I would recommend doing that. The obvious disadvantage is that it becomes user responsibility to make sure that things are consistent and that the config that was used for submission is reproducible.

Potentially something for another PR.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: If I understand correctly, this is superfluous. @gaswk you presumably cloned your fork from `CLDConfig` to `afs`, right? Then the job should be able to access it without having to clone it again if I am not mistaken:
"This has access to two shared file systems, AFS and EOS. The batch system can also see both filesystems."([batchdocs](https://batchdocs.web.cern.ch/concepts/dataflow.html))

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py:212]: Another option is to outsource the execution and move it to a new script. This script can call the `main` function of `condorJobs_sim` and then submit the jobs.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (nitasad) [reply TrackingPerformance/Condor/condorJobs_sim.py]: From what I can see, there doesn't seem to be much change compared to what I did. This check will rerun the simulation for the missing files because the jobs crashed or encountered other issues, which happens quite often. Therefore, you need to use the same variables. It will also verify if the output files contain the correct number of events. I've encountered this issue several times where the files exist but contain only a few events.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (nitasad) [reply TrackingPerformance/Condor/condorJobs_sim.py]: Yes, it has access to `afs` and can use your own fork from there. However, I think it is preferable to use the steering file from `https://github.com/key4hep/` by default. We could also make it configurable, allowing you to specify a different steering file if desired?

Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: In `cld_config` line 26, one can specify the steering file:
```
sim_steering_file = detectorDIR / "CHANGE" / "PATH"  # FIXME
```
The one from `https://github.com/key4hep/` should be accessible via `cvmfs`, shouldn't it?

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: To me, it seems like the `need_to_create_scripts` variable is only set to `True` if the condition in line `161` equals `False`. The if statement `events_tree.GetEntries() == int(Nevt_per_job)` misses an `else` to ensure that missing events are produced. I will create an issue based on it to keep this pull request cleaner

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [reply TrackingPerformance/Condor/condorJobs_sim.py]: > The one from `https://github.com/key4hep/` should be accessible via `cvmfs`, shouldn't it?

It should be accessible in the future but not at the moment. The ILD equivalent [`ddsim_steer.py`](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/ddsim_steer.py) is also not available (yet) on cvmfs as far as I know...

Comment (tmadlener) [reply TrackingPerformance/Condor/condorJobs_sim.py]: Technically the tags are available under `/cvmfs/ilc.desy.de/sw/ILDConfig/vxx-yy-zz/`, but not within a key4hep release yet.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: done

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: I removed the cloning. If you want another solution, I suggest to open a new issue

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py:212]: Moved to separate issue

**Comment by Zehvogel** [comment]
@Victor-Schwan is this ready now?

**Comment by Victor-Schwan** [comment]
Not ready yet. When I ran it on `lxbatch` there was a permission problem in `sim_jobs` which needs to be debugged. And I have an almost completed commit waiting removing double definitions of vars and some other refacs

**Comment by Zehvogel** [comment]
Ok

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_reco.py:124]: I think having a directory per particle type is useful :)

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_reco.py]: I know this is from Gaelle but on the one hand I am completely confused what is supposed to hapen here syntax wise, while on the other hand we should make the whole `ResVDX_UV_` optional (also below).

It would be nice to be able to specify additional arguments to k4run in the config and then this could go there.

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Zehvogel) [inline TrackingPerformance/Condor/condorJobs_sim.py]: :warning:

**Review by Zehvogel — COMMENTED** [review]
Some more comments.

I guess we will need a future PR on `simJobs_reco` to make it a bit more flexible but for now I am happy.

I will merge after @gaswk approves. :)

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_reco.py]: Moved to a new issue

File: TrackingPerformance/Condor/condorJobs_sim.py
Comment (Victor-Schwan) [reply TrackingPerformance/Condor/condorJobs_sim.py]: `det_mod_paths` key is in the next commit chosen from `config.detector_model_list[0]`

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (nitasad) [inline TrackingPerformance/Condor/condorJobs_reco.py]: Should be `.edm4hep.root` to be consistent

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (nitasad) [inline TrackingPerformance/Condor/condorJobs_reco.py]: `input_file` and `config.N_EVTS_PER_JOB` need to be of type str

File: TrackingPerformance/Condor/condorJobs_reco.py
Comment (nitasad) [inline TrackingPerformance/Condor/condorJobs_reco.py]: Please use https://github.com/key4hep/CLDConfig.git and I think `ResVDX_UV_` argument should be removed, at least for now

**Comment by Zehvogel** [comment]
@gaswk is everything fine from your side now?

**Comment by nitasad** [comment]
> @gaswk is everything fine from your side now?

I am waiting for your response to my previous comment regarding the naming conventions for the output files. To maintain consistency, should we use `_` instead of `.` for all output names? Currently, the outputs from the reco are named `_edm4hep.root` and `_aida.root` by default.

**Comment by Zehvogel** [comment]
> I am waiting for your response to my previous comment regarding the naming conventions for the output files. To maintain consistency, should we use `_` instead of `.` for all output names? Currently, the outputs from the reco are named `_edm4hep.root` and `_aida.root` by default.

Oops, sorry for missing that. I would prefer `.` instead of `_` but if `CLDConfig` still produces `_` we would need to change it there first of course :/ I think a have a change for that ready somewhere but that also changes some more parts.

@Victor-Schwan opposite to previous instructions I would then suggest that you change back to `_` for this PR and we make the change to `.` later...

**Comment by Victor-Schwan** [comment]
I added a bool var in the config file which switches between `_edm4hep.root` and `.edm4hep.root` paths. What do you think? @Zehvogel @gaswk

File: TrackingPerformance/config_template.py
Comment (Zehvogel) [inline TrackingPerformance/config_template.py]:
```suggestion
# True: filepath_edm4hep.root; False: filepath.edm4hep.root
EDM4HEP_SUFFIX_WITH_UNDERSCORE = False
```

File: TrackingPerformance/config_v.py
Comment (Zehvogel) [inline TrackingPerformance/config_v.py]:
```suggestion
# True: filepath_edm4hep.root; False: filepath.edm4hep.root
```

**Review by Zehvogel — CHANGES_REQUESTED** [review]
I think you meant this?

File: TrackingPerformance/config_template.py
Comment (Victor-Schwan) [reply TrackingPerformance/config_template.py]: yes, thanks

**Review by Zehvogel — APPROVED** [review]

**Comment by nitasad** [comment]
Everything is fine for me now!

**Pull request merged**
